egs/wham/MixIT/train.py

```diff
@@ -0,0 +1,180 @@
+import os
+import argparse
+import json
+
+import torch
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+from torch.utils.data import DataLoader
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import ModelCheckpoint, EarlyStopping
+
+from asteroid import DPRNNTasNet
+from asteroid.data.wham_dataset import WhamDataset
+from asteroid.engine.optimizers import make_optimizer
+from asteroid.engine.system import System
+from asteroid.losses import PITLossWrapper, pairwise_neg_sisdr, MixITLossWrapper
+
+# Keys which are not in the conf.yml file can be added here.
+# In the hierarchical dictionary created when parsing, the key `key` can be
+# found at dic['main_args'][key]
+
+# By default train.py will use all available GPUs. The `id` option in run.sh
+# will limit the number of available GPUs for train.py .
+parser = argparse.ArgumentParser()
+parser.add_argument("--exp_dir", default="exp/tmp", help="Full path to save best validation model")
+
+
+class MixITSystem(System):
+    def training_step(self, batch, batch_nb):
+        mixtures, oracle = batch
+
+        bsz = mixtures.shape[0]
+        mix1 = mixtures[bsz // 2 :]
+        mix2 = mixtures[: bsz // 2]
+        moms = mix1 + mix2
+
+        est_sources = self(moms)
+        loss = self.loss_func["mixit"](est_sources, torch.stack((mix1, mix2)).transpose(0, 1))
+
+        with torch.no_grad():
+            # we cheat during training and check oracle performance ^^
+            loss_oracle = self.loss_func["pit"](
+                est_sources, torch.cat((oracle[bsz // 2 :], oracle[: bsz // 2]), 1)
+            )
```

Comment:
Actually, `MixITWrapper` could be used as `PITLossWrapper` when the number of sources is the same as the number of mixtures. Maybe we should make that, it would be a bit cleaner.

Reply:
Also on validation then. But I guess it could be also confusing for someone. That is why we choose to have two separate classes even if MixIT is a generalization of PIT

Reply:
Haha, yes. I don't know then..